feat: move to accelerate launch for distributed training

kmehant opened the pull request:
<!-- Thank you for the contribution! -->

### Description of the change

<!-- Please summarize the changes -->

### Related issue number

Closes #87 

### How to verify the PR

I have launched a multi gpu training using FSDP and accelerate

```bash
accelerate launch --main_process_ip $MASTER_ADDR --main_process_port $MASTER_PORT --config_file config/accelerate_fsdp_config.yaml tuning/sft_trainer.py --peft_method none --tokenizer_name_or_path $MODEL_PATH --model_name_or_path ${MODEL_PATH} --data_path ${DATA_PATH} --output_dir ${OUTPUT_PATH} --num_train_epochs 5 --per_device_train_batch_size 2 --per_device_eval_batch_size 4 --evaluation_strategy "no" --save_strategy "epoch" --learning_rate "1e-5" --warmup_ratio 0.03 --lr_scheduler_type "cosine" --logging_steps 1 --include_tokens_per_second --packing False --response_template "\n### Label:" --dataset_text_field "output" --use_flash_attn False --torch_dtype bfloat16
```
`$MASTER_ADDR` and `$MASTER_PORT` are set by the job launcher (Pytorch Job launcher from kubeflow operator). I have used `twitter_complaints.json` for the training data and Llama 7B as the base model. Rest of the training arguments used are shown in the above command.

I have used the accelerate config thats being added by this PR (config/accelerate_fsdp_llama_2_procs.yaml).

Some screenshots on the training.

#### GPU and memory utilization
<img width="567" alt="Screenshot 2024-03-13 at 1 44 39 PM" src="https://github.com/foundation-model-stack/fms-hf-tuning/assets/15800200/16a9540e-621e-4d47-8951-d55d0d9860ff">
<img width="707" alt="Screenshot 2024-03-13 at 1 58 49 PM" src="https://github.com/foundation-model-stack/fms-hf-tuning/assets/15800200/46e1fa9e-8df9-4e0a-9bd4-ed3d9e81b1f5">
<img width="1418" alt="Screenshot 2024-03-13 at 1 58 58 PM" src="https://github.com/foundation-model-stack/fms-hf-tuning/assets/15800200/c3a38dd4-77f9-4e9b-9751-7c7be0fd4824">

#### Training loss curve
<img width="707" alt="Screenshot 2024-03-13 at 1 56 14 PM" src="https://github.com/foundation-model-stack/fms-hf-tuning/assets/15800200/d5554f9f-01a1-417a-b3d5-02776c2200b1">


## Concerns

Saving model

ref - https://huggingface.co/blog/ram-efficient-pytorch-fsdp

set full state dict before saving the model
```
    trainer.accelerator.state.fsdp_plugin.set_state_dict_type("FULL_STATE_DICT")
```


### Was the PR tested

<!-- Describe how PR was tested -->
- [ ] I have added >=1 unit test(s) for every new method I have added.
- [ ] I have ensured all unit tests pass

Ssukriti commented on config/accelerate_fsdp_llama_2_procs.yaml:
```diff
@@ -0,0 +1,27 @@
+  fsdp_sharding_strategy: FULL_SHARD
+  fsdp_state_dict_type: SHARDED_STATE_DICT
+  fsdp_sync_module_states: true
+  fsdp_transformer_layer_cls_to_wrap: LlamaDecoderLayer
```
hmm I think its safest to unset it by default like Fabian did in his PR https://github.com/foundation-model-stack/fms-hf-tuning/pull/91/files#diff-7e2cb0844b4ef444122dc5020b67bfd949b38de258a1c63dbfa8abad0a746988R37

kmehant replied:
@Ssukriti  Thank you, I have cherry picked this from Fabian's PR automatically credited him in this commit - https://github.com/foundation-model-stack/fms-hf-tuning/pull/92/commits/4aa0153ab0c55ac5b305b8d60e8f00f89cc354d4

Ssukriti commented on README.md:
```diff
@@ -83,11 +87,16 @@ python tuning/sft_trainer.py  \
+MASTER_ADDR=x.x.x.x # The IP addresss of the node with rank 0
+
+accelerate launch --main_process_ip $MASTER_ADDR --main_process_port $MASTER_PORT \
+--config_file config/accelerate_fsdp_llama_2_procs.yaml \
```
does this change also need to be added? https://github.com/foundation-model-stack/fms-hf-tuning/pull/91/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R105

kmehant replied:
The previous `--bf16` would fire up mixed precision training whereas `--torch_dtype` would load the model weights in the given dtype. We can use both in conjunction. However, I have moved on with Fabian's change as its much more simple without bringing in mixed precision training (using --bf16) and loading the model in dtype bfloat16 which is typical (even I have used this in my command as shown in the PR description (https://github.com/foundation-model-stack/fms-hf-tuning/pull/92#issue-2183504285)) This is part of the cherry picked commit and credited him :)

Ssukriti commented on README.md:
````diff
@@ -83,11 +87,16 @@ python tuning/sft_trainer.py  \
 ```
 
 ### Multiple GPUs with FSDP
+
````
maybe taking lines 87-93 from here is a good idea, since fabian added them https://github.com/foundation-model-stack/fms-hf-tuning/pull/91/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R87

kmehant replied:
Sure @Ssukriti, I have cherry picked this from Fabian's PR automatically credited him in this commit - https://github.com/foundation-model-stack/fms-hf-tuning/commit/4aa0153ab0c55ac5b305b8d60e8f00f89cc354d4

Ssukriti reviewed (changes requested):
thank for the PR!

Minor suggestions , unfortunately we had a duplicate created at same time https://github.com/foundation-model-stack/fms-hf-tuning/pull/91/ 

Fabian had also volunteered earlier to add it, but I thought he was busy and then you both crated PR at same time.

To keep things fair, do you mind co-commiting so you both get credit? You could rebase commits or open a new PR

I reviewed both your PRs and added content here that was different and could be useful.

kmehant commented:
@Ssukriti I have addressed your comments and cherry picked Fabian's commit from his PR to here (automatically credits him at commit https://github.com/foundation-model-stack/fms-hf-tuning/pull/92/commits/4aa0153ab0c55ac5b305b8d60e8f00f89cc354d4). Looking forward to your review, thanks.

fabianlim commented on README.md:
```diff
@@ -83,15 +87,38 @@ python tuning/sft_trainer.py  \
+--config_file fixtures/accelerate_fsdp_defaults.yaml \
+--num_machines=1 \
+--num_processes=8 \ 
+--main_process_port=1234 \
```
```suggestion
--main_process_port=$MASTER_PORT \
```
since we had already defined `MASTER_PORT` above we should use it./

fabianlim commented on README.md:
```diff
@@ -60,6 +60,10 @@ Current supported and tested models are `Llama2` (7 and 13B configurations have
 # if you want to use one GPU on multi-gpu machine
 export CUDA_VISIBLE_DEVICES=0
 
+MODEL_PATH=llama-7b-hf # Huggingface model id or path to a checkpoint
+DATA_PATH=twitter_complaints.json # Path to the dataset
+OUTPUT_PATH=out # Path to the output folder where the checkpoints are saved
+
```
```suggestion
```
suggest to now have these extra arguments here, since this README is for generic use case

also `llama-7b-hf` is not the official huggingface name actually,

Ssukriti replied:
If your intent is to avoid users from directly executing as it is, and force them to read and set variables themselves, we can comment out the values like below, instead of deleting:

```
# Please set the environment variables:
# MODEL_PATH = llama-7b-hf # Huggingface model id or path to a checkpoint
# TRAIN_DATA_PATH=twitter_complaints.json # Path to the train dataset
# OUTPUT_PATH=out # Path to the output folder where the checkpoints are saved
```
I think its good to have it in README though, as its just an example demonstrating that we need to set these environment variables and roughly what their values look like. Users trying out the repository may not understand `data_path ` accepts a single json for example - we have explained how data should be formatted directly above in README, but good to show example that it accepts a single JSON

fabianlim replied:
I agree its in the prompt tuning one also. If this is fixed, it needs to be fixed in both places. I leave it to @Ssukriti to decide which version she wants.

Ssukriti replied:
You can change the name of the model as you want , I just meant to include it as example . You can use the meta one

kmehant replied:
Updated @Ssukriti !

fabianlim commented on README.md:
```diff
@@ -104,16 +131,14 @@ tuning/sft_trainer.py \
 
 
-For `GPTBigCode` models, Hugging Face has enabled Flash v2 and one can simply replace the `'LlamaDecoderLayer'` with `'GPTBigCodeBlock'` in `tuning/config/fsdp_config.json` for proper sharding of the model.
+Typically the transformer module is passed to form FSDP unit. For `GPTBigCode` models, Hugging Face has enabled Flash v2 and one can simply replace the `'LlamaDecoderLayer'` with `'GPTBigCodeBlock'` in `config/accelerate_fsdp_llama_2_procs.yaml` for proper sharding of the model.
```
```suggestion
```
Feel that this text is not really needed. and we should remove `config/accelerate_fsdp_llama_2_procs.yaml`, since the new file `fixtures/accelerate_fsdp_defaults.yaml` can run Llama with 2 processes, by simply augmenting the command `--num_processes=2` to `accelerate launch`.

fabianlim commented on README.md:
````diff
@@ -83,15 +87,38 @@ python tuning/sft_trainer.py  \
+MASTER_ADDR=x.x.x.x # The IP addresss of the node with rank 0
+
+
+```bash
+accelerate launch \
+--main_process_ip $MASTER_ADDR \
+--main_process_port $MASTER_PORT \
+--config_file fixtures/accelerate_fsdp_defaults.yaml \
+--num_machines=1 \
````
```suggestion
```bash
accelerate launch \
--main_process_port $MASTER_PORT \
--config_file fixtures/accelerate_fsdp_defaults.yaml \
```

`--num_machines 1` is already default in `accelerate_fsdp_defaults.yaml`. Also as mentioned above I feel better to not discuss `--main_process_ip $MASTER_ADDR`

fabianlim commented on README.md:
```diff
@@ -83,15 +87,38 @@ python tuning/sft_trainer.py  \
---nnodes=1 \
---nproc_per_node=8 \ 
---master_port=1234 \
+MODEL_PATH=llama-7b-hf # Huggingface model id or path to a checkpoint
+DATA_PATH=twitter_complaints.json # Path to the dataset
+OUTPUT_PATH=out # Path to the output folder where the checkpoints are saved
+
+# MASTER_PORT and MASTER_ADDR are essential for multi node training and 
+# not needed for multi gpu in single node
+MASTER_PORT=1234 # The port at which the process with rank 0 listens to
+MASTER_ADDR=x.x.x.x # The IP addresss of the node with rank 0
```
```suggestion
# MASTER_PORT should be set to an unused port
MASTER_PORT=1234 
```

For now I suggest to keep things more streamlined and not discuss about multi node for now

Ssukriti replied:
yes agreed multinode is not on our roadmap, so good to avoid. Good catch Fabian, thank you!

fabianlim commented on examples/prompt_tuning_twitter_complaints/README.md:
```diff
@@ -35,15 +35,21 @@ MODEL_PATH=llama-7b-hf
---nnodes=1 \
---nproc_per_node=8  \
---master_port=1234  \
+# MASTER_PORT and MASTER_ADDR are essential for multi node training and 
+# not needed for multi gpu in single node
+MASTER_PORT=1234 # The port at which the process with rank 0 listens to
+MASTER_ADDR=x.x.x.x # The IP addresss of the node with rank 0
+
+accelerate launch \
+--main_process_ip $MASTER_ADDR \
+--main_process_port $MASTER_PORT \
+--config_file fixtures/accelerate_fsdp_defaults.yaml \
```
```suggestion
# MASTER_PORT should be set to an unused port
MASTER_PORT=1234 # The port at which the process with rank 0 listens to

accelerate launch \
--main_process_port $MASTER_PORT \
--config_file fixtures/accelerate_fsdp_defaults.yaml \
```

fabianlim reviewed (commented):
@kmehant @Ssukriti I have some suggestions. I feel the PR can be further streamlined.

Ssukriti commented on README.md:
```diff
@@ -83,15 +87,38 @@ python tuning/sft_trainer.py  \
 --model_name_or_path $MODEL_PATH \
---training_data_path $DATA_PATH \
---bf16 True \
+--data_path $DATA_PATH \
```
this is now training_data_path on latest main https://github.com/foundation-model-stack/fms-hf-tuning/blob/main/tuning/config/configs.py#L45 . Please use this value and name of environment variable can also be updated.

This was done to distinguish between train and validation datasets

Ssukriti reviewed (commented):
will approve and merge when Fabian's comments are addressed. I had another small comment on the name of `training_data_path` variable.

Just on the environment variables , I would suggest keeping them as examples to avoid confusion on how to set them. 
`# Please set the environment variables:
# MODEL_PATH = llama-7b-hf # Huggingface model id or path to a checkpoint
# TRAIN_DATA_PATH=twitter_complaints.json # Path to the train dataset
# OUTPUT_PATH=out # Path to the output folder where the checkpoints are saved`

Will wait for other comments from Fabian to be addressed. Thank you!

kmehant commented:
@Ssukriti @fabianlim I have collated all your review comments and have updated the PR, please review, thank you.

kmehant commented:
@Ssukriti The PR is ready, thank you.

Ssukriti commented on README.md:
```diff
@@ -60,6 +60,10 @@ Current supported and tested models are `Llama2` (7 and 13B configurations have
+
 python tuning/sft_trainer.py  \
 --model_name_or_path $MODEL_PATH  \
 --training_data_path $DATA_PATH  \
```
```suggestion
--training_data_path $TRAIN_DATA_PATH  \
```

Ssukriti commented on README.md:
```diff
@@ -60,6 +60,10 @@ Current supported and tested models are `Llama2` (7 and 13B configurations have
 # if you want to use one GPU on multi-gpu machine
 export CUDA_VISIBLE_DEVICES=0
 
+MODEL_PATH=llama-7b-hf # Huggingface model id or path to a checkpoint
```
```suggestion
# MODEL_PATH=meta-llama/Llama-2-7b-hf # Huggingface model id or path to a checkpoint
```

Ssukriti commented on README.md:
```diff
@@ -60,6 +60,10 @@ Current supported and tested models are `Llama2` (7 and 13B configurations have
 export CUDA_VISIBLE_DEVICES=0
 
+MODEL_PATH=llama-7b-hf # Huggingface model id or path to a checkpoint
+DATA_PATH=twitter_complaints.json # Path to the dataset
```
```suggestion
# TRAIN_DATA_PATH=twitter_complaints.json # Path to the dataset
```

Ssukriti commented on README.md:
```diff
@@ -60,6 +60,10 @@ Current supported and tested models are `Llama2` (7 and 13B configurations have
 
+MODEL_PATH=llama-7b-hf # Huggingface model id or path to a checkpoint
+DATA_PATH=twitter_complaints.json # Path to the dataset
+OUTPUT_PATH=out # Path to the output folder where the checkpoints are saved
```
```suggestion
# OUTPUT_PATH=out # Path to the output folder where the checkpoints are saved
```

Ssukriti commented on examples/prompt_tuning_twitter_complaints/README.md:
```diff
@@ -30,20 +30,22 @@ dataset.to_json("twitter_complaints.json")
+--config_file fixtures/accelerate_fsdp_defaults.yaml \
 tuning/sft_trainer.py  \
 --model_name_or_path $MODEL_PATH  \
 --training_data_path $DATA_PATH  \
```
```suggestion
--training_data_path $TRAIN_DATA_PATH  \
```

Ssukriti commented on README.md:
```diff
@@ -83,15 +87,33 @@ python tuning/sft_trainer.py  \
+--main_process_port=$MASTER_PORT \
 tuning/sft_trainer.py \
 --model_name_or_path $MODEL_PATH \
 --training_data_path $DATA_PATH \
```
```suggestion
--training_data_path $TRAIN_DATA_PATH \
```

Ssukriti reviewed (approved):
thank you!
